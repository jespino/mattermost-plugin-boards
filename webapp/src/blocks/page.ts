// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Block, createBlock} from './block'

type PageFields = {
    contentOrder: Array<string | string[]>
    properties: Record<string, any>
    icon: string
}

type Page = Block & {
    fields: PageFields
}

function createPage(block?: Block): Page {
    const contentOrder: Array<string|string[]> = []
    const contentIds = block?.fields?.contentOrder?.filter((id: any) => id !== null)

    if (contentIds?.length > 0) {
        for (const contentId of contentIds) {
            if (typeof contentId === 'string') {
                contentOrder.push(contentId)
            } else {
                contentOrder.push(contentId.slice())
            }
        }
    }

    return {
        ...createBlock(block),
        type: 'page',
        fields: {
            contentOrder,
            icon: block?.fields?.icon,
            properties: {...block?.fields?.properties},
        },
    }
}

function sortPagesAlphabetically(pages: Page[]): Page[] {
    // Strip leading emoji to prevent unintuitive results
    return pages.map((v) => {
        return {page: v, title: v.title.replace(/^\p{Emoji}*\s*/u, '')}
    }).sort((v1, v2) => v1.title.localeCompare(v2.title)).map((v) => v.page)
}

export {Page, sortPagesAlphabetically, createPage}
