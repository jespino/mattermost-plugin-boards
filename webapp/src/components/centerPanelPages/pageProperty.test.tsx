// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import '@testing-library/jest-dom'
import {act, render} from '@testing-library/react'

import React from 'react'
import {Provider as ReduxProvider} from 'react-redux'

import {IPropertyTemplate} from '../../blocks/board'
import {TestBlockFactory} from '../../test/testBlockFactory'
import {mockDOM, mockStateStore, wrapIntl} from '../../testUtils'

import PageProperty from './pageProperty'

beforeAll(() => {
    mockDOM()
})

describe('components/pageProperty', () => {
    const board = TestBlockFactory.createBoard()
    board.id = 'test-id'
    const fakeProperty: IPropertyTemplate = {id: 'test-prop', name: 'Test prop', type: 'text', options: []}
    const state = {
        users: {
            boardUsers: {
                1: {username: 'abc'},
                2: {username: 'd'},
                3: {username: 'e'},
                4: {username: 'f'},
                5: {username: 'g'},
            },
        },
        teams: {
            current: {id: 'team-id'},
        },
        boards: {
            current: board.id,
            boards: {
                [board.id]: board,
            },
            myBoardMemberships: {
                [board.id]: {userId: 'user_id_1', schemeAdmin: true},
            },
        },
        clientConfig: {
            value: {},
        },
    }
    const store = mockStateStore([], state)

    beforeEach(() => {
        jest.clearAllMocks()
    })

    test('should match snapshot without value', async () => {
        let container
        const page = TestBlockFactory.createPage(board)
        await act(async () => {
            const result = render(wrapIntl(
                <ReduxProvider store={store}>
                    <PageProperty
                        page={page}
                        board={board}
                        readonly={false}
                        canEditBoardProperties={true}
                        canEditBoardCards={true}
                        newTemplateId={''}
                        propertyTemplate={fakeProperty}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot with value', async () => {
        let container
        const page = TestBlockFactory.createPage(board)
        page.fields.properties = {'test-prop': 'Test value'}
        await act(async () => {
            const result = render(wrapIntl(
                <ReduxProvider store={store}>
                    <PageProperty
                        page={page}
                        board={board}
                        readonly={false}
                        canEditBoardProperties={true}
                        canEditBoardCards={true}
                        newTemplateId={''}
                        propertyTemplate={fakeProperty}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot readonly without value', async () => {
        let container
        const page = TestBlockFactory.createPage(board)
        await act(async () => {
            const result = render(wrapIntl(
                <ReduxProvider store={store}>
                    <PageProperty
                        page={page}
                        board={board}
                        readonly={true}
                        canEditBoardProperties={true}
                        canEditBoardCards={true}
                        newTemplateId={''}
                        propertyTemplate={fakeProperty}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot readonly with value', async () => {
        let container
        const page = TestBlockFactory.createPage(board)
        page.fields.properties = {'test-prop': 'Test value'}
        await act(async () => {
            const result = render(wrapIntl(
                <ReduxProvider store={store}>
                    <PageProperty
                        page={page}
                        board={board}
                        readonly={true}
                        canEditBoardProperties={true}
                        canEditBoardCards={true}
                        newTemplateId={''}
                        propertyTemplate={fakeProperty}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot without edit board permissions', async () => {
        let container
        const page = TestBlockFactory.createPage(board)
        page.fields.properties = {'test-prop': 'Test value'}
        await act(async () => {
            const result = render(wrapIntl(
                <ReduxProvider store={store}>
                    <PageProperty
                        page={page}
                        board={board}
                        readonly={false}
                        canEditBoardProperties={false}
                        canEditBoardCards={true}
                        newTemplateId={''}
                        propertyTemplate={fakeProperty}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot without edit card permissions', async () => {
        let container
        const page = TestBlockFactory.createPage(board)
        page.fields.properties = {'test-prop': 'Test value'}
        await act(async () => {
            const result = render(wrapIntl(
                <ReduxProvider store={store}>
                    <PageProperty
                        page={page}
                        board={board}
                        readonly={false}
                        canEditBoardProperties={true}
                        canEditBoardCards={false}
                        newTemplateId={''}
                        propertyTemplate={fakeProperty}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })
})
