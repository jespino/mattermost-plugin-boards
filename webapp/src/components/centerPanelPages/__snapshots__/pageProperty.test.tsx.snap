// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pageProperty should match snapshot readonly with value 1`] = `
<div>
  <div
    class="PageProperty octo-propertyrow"
  >
    <div
      class="octo-propertyname octo-propertyname--readonly"
    >
      Test prop
    </div>
    <div
      class="octo-propertyvalue octo-propertyvalue--readonly"
    >
      Test value
    </div>
  </div>
</div>
`;

exports[`components/pageProperty should match snapshot readonly without value 1`] = `
<div>
  <div
    class="PageProperty octo-propertyrow"
  >
    <div
      class="octo-propertyname octo-propertyname--readonly"
    >
      Test prop
    </div>
    <div
      class="octo-propertyvalue octo-propertyvalue--readonly"
    />
  </div>
</div>
`;

exports[`components/pageProperty should match snapshot with value 1`] = `
<div>
  <div
    class="PageProperty octo-propertyrow"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <div
        class="octo-propertyname"
      >
        <button
          class="Button"
          type="button"
        >
          <span>
            Test prop
          </span>
        </button>
      </div>
    </div>
    <input
      class="Editable octo-propertyvalue"
      placeholder="Empty"
      spellcheck="true"
      style="width: 100%;"
      title="Test value"
      value="Test value"
    />
  </div>
</div>
`;

exports[`components/pageProperty should match snapshot without edit board permissions 1`] = `
<div>
  <div
    class="PageProperty octo-propertyrow"
  >
    <div
      class="octo-propertyname octo-propertyname--readonly"
    >
      Test prop
    </div>
    <input
      class="Editable octo-propertyvalue"
      placeholder="Empty"
      spellcheck="true"
      style="width: 100%;"
      title="Test value"
      value="Test value"
    />
  </div>
</div>
`;

exports[`components/pageProperty should match snapshot without edit card permissions 1`] = `
<div>
  <div
    class="PageProperty octo-propertyrow"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <div
        class="octo-propertyname"
      >
        <button
          class="Button"
          type="button"
        >
          <span>
            Test prop
          </span>
        </button>
      </div>
    </div>
    <div
      class="octo-propertyvalue octo-propertyvalue--readonly"
    >
      Test value
    </div>
  </div>
</div>
`;

exports[`components/pageProperty should match snapshot without value 1`] = `
<div>
  <div
    class="PageProperty octo-propertyrow"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <div
        class="octo-propertyname"
      >
        <button
          class="Button"
          type="button"
        >
          <span>
            Test prop
          </span>
        </button>
      </div>
    </div>
    <input
      class="Editable octo-propertyvalue"
      placeholder="Empty"
      spellcheck="true"
      style="width: 100%;"
      title=""
      value=""
    />
  </div>
</div>
`;
